ocaml/runtime/extern.c

```diff
@@ -1157,45 +1206,140 @@ CAMLprim value caml_obj_reachable_words(value v)
          between major heap blocks and out-of-heap blocks,
          and the test above is always false,
          so we end up counting out-of-heap blocks too. */
-    } else if (extern_lookup_position(v, &pos, &h)) {
-      /* Already seen and counted, nothing to do */
     } else {
-      header_t hd = Hd_val(v);
-      tag_t tag = Tag_hd(hd);
-      mlsize_t sz = Wosize_hd(hd);
-      /* Infix pointer: go back to containing closure */
-      if (tag == Infix_tag) {
-        v = v - Infix_offset_hd(hd);
-        continue;
+      previously_marked = extern_lookup_position(v, &mark, &h);
+      if (!previously_marked) {
+        /* Invariant: v != root as we have marked roots before running this function.
+         * So we can safely assign new_mark to identifier */
+        should_traverse = 1;
+        new_mark = identifier;
+      } else if (mark == RootUnprocessed && v == root) {
+        should_traverse = 1;
+        new_mark = RootProcessed;
+      } else if (mark == Shared || mark == RootUnprocessed || mark == RootProcessed) {
+        should_traverse = 0;
+      } else if (mark == identifier) {
+        should_traverse = 0;
+      } else {
+        /* mark is some other root's identifier */
+        should_traverse = 1;
+        new_mark = Shared;
       }
-      /* Remember that we've visited this block */
-      extern_record_location(v, h);
-      /* The block contributes to the total size */
-      size += 1 + sz;           /* header word included */
-      if (tag < No_scan_tag) {
-        /* i is the position of the first field to traverse recursively */
-        uintnat i =
-          tag == Closure_tag ? Start_env_closinfo(Closinfo_val(v)) : 0;
-        if (i < sz) {
-          if (i < sz - 1) {
-            /* Remember that we need to count fields i + 1 ... sz - 1 */
-            sp++;
-            if (sp >= extern_stack_limit) sp = extern_resize_stack(sp);
-            sp->v = &Field(v, i + 1);
-            sp->count = sz - i - 1;
-          }
-          /* Continue with field i */
-          v = Field(v, i);
+
+      if (should_traverse) {
+        header_t hd = Hd_val(v);
+        tag_t tag = Tag_hd(hd);
+        mlsize_t sz = Wosize_hd(hd);
+        /* Infix pointer: go back to containing closure */
```

Comment:
I think this is slightly wrong: the check for Infix_tag going back to the containing closure should be before `extern_lookup_position`. (All Infix_tag headers on the same closure are the same block, so should share a mark. So we need to normalise _before_ lookup)

Reply:
For reference, this is the same behavior as the [original code](https://github.com/ocaml-flambda/flambda-backend/blob/caadaffa7979739d1f69a8595a1ffe5a3cbd0a35/ocaml/runtime/extern.c#L1139).

I think this might be fine because of the `continue` since it means we will never record the `Infix_tag` block in the hashmap.
I'm happy to move this earlier in the function though for clarity's sake.